library/core/src/iter/traits/iterator.rs

```diff
@@ -1777,6 +1777,62 @@ pub trait Iterator {
         FromIterator::from_iter(self)
     }
 
+    /// Fallibly transforms an iterator into a collection, short circuiting if
+    /// a failure is encountered while collecting.
+    ///
+    /// `try_collect()` is essentially a wrapper around [`collect()`][`collect`] for types
+    /// that can already be fallibly constructed from iterators. Its main use case
+    /// is simplifying the conversion from iterators of [`Option<T>`][`Option`] into `Option<Vec<T>>`
+    /// (a similar technique exists for [`Result`]).
+    ///
+    /// In fact, `try_collect()` is generic over types that implement [`Try`], so any such type
+    /// can be fallibly collected into its output type in the same way.
+    ///
+    /// # Examples
+    /// Successfully collecting an iterator of `Option<i32>` into `Option<Vec<i32>>`:
+    /// ```
+    /// #![feature(iterator_try_collect)]
+    ///
+    /// let u = vec![Some(1), Some(2), Some(3)];
+    /// let v = u.into_iter().try_collect::<Vec<i32>>();
+    /// assert_eq!(v, Some(vec![1, 2, 3]));
+    /// ```
+    ///
+    /// Failing to collect in the same way:
+    /// ```
+    /// #![feature(iterator_try_collect)]
+    ///
+    /// let u = vec![Some(1), Some(2), None, Some(3)];
+    /// let v = u.into_iter().try_collect::<Vec<i32>>();
+    /// assert_eq!(v, None);
+    /// ```
+    ///
+    /// A similar example, but with `Result`:
+    /// ```
+    /// #![feature(iterator_try_collect)]
+    ///
+    /// let u: Vec<Result<i32, ()>> = vec![Ok(1), Ok(2), Ok(3)];
+    /// let v = u.into_iter().try_collect::<Vec<i32>>();
+    /// assert_eq!(v, Ok(vec![1, 2, 3]));
+    ///
+    /// let u = vec![Ok(1), Ok(2), Err(()), Ok(3)];
+    /// let v = u.into_iter().try_collect::<Vec<i32>>();
+    /// assert_eq!(v, Err(()));
+    /// ```
+    ///
+    /// [`collect`]: Iterator::collect
+    #[inline]
+    #[unstable(feature = "iterator_try_collect", issue = "none")]
+    fn try_collect<T>(&mut self) -> ChangeOutputType<Self::Item, T>
+    where
+        Self: Sized,
+        <Self as Iterator>::Item: Try,
+        <<Self as Iterator>::Item as Try>::Residual: Residual<T>,
+        ChangeOutputType<Self::Item, T>: FromIterator<Self::Item>,
```

Comment:
Hmm, personally I wasn't thinking of having this go through `FromIterator` for `Result` at all, since that would mean that people still need to *write* that impl.

If this were to go through `iter::try_process` instead, for example, then it wouldn't need `Result: FromIterator`

https://github.com/rust-lang/rust/blob/393fdc10483da930cdbb00eabc3635030d2e776f/library/core/src/iter/adapters/mod.rs#L140-L156

(That might not work exactly as-is, though.  It might need a new `try_process_mut(&mut I, F)` version so that continuing after starting works out well, which might mean changing `GenericShunt` to also hold `&mut 'a I` instead of just `I`.)

But regardless of the precise implementation details, I'd suggest trying to get this to work:
```rust
    use std::ops::ControlFlow::{Break, Continue};
    let a = [Continue(1), Continue(2), Break(3), Continue(4), Continue(5)];
    let mut it = a.into_iter();
    let r = it.try_collect::<Vec<_>>();
    assert_eq!(r, Break(3));
    let r = it.try_collect::<Vec<_>>();
    assert_eq!(r, Continue(vec![4, 5]));
```
It might need some more type hints in the asserts, but otherwise it should work, I think.  And it hits the core things here: a `Try`-but-not-`FromIterator` type, the resumption behaviour, and the turbofishing being only the collection.

Reply:
Ok yeah, I see what you're saying. I think I have a better grasp of the main points of the feature now, so I'll modify my implementation per your comments. Thanks!